ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl/Parser.hs

```diff
@@ -0,0 +1,129 @@
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving  #-}
+module Ouroboros.Consensus.Storage.VolatileDB.Impl.Parser (
+    Parser (..)
+  , ParsedBlockInfo (..)
+  , ParseError (..)
+  , blockFileParser
+    -- * Auxiliary
+  , extractBlockInfo
+  ) where
+
+import           Data.Bifunctor (bimap)
+import qualified Data.ByteString.Lazy as Lazy
+import           Data.Word (Word64)
+import           Streaming.Prelude (Of (..), Stream)
+import qualified Streaming.Prelude as S
+
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.Util.CBOR (ReadIncrementalErr (..),
+                     withStreamIncrementalOffsets)
+import           Ouroboros.Consensus.Util.IOLike
+
+import           Ouroboros.Consensus.Storage.ChainDB.Serialisation
+import           Ouroboros.Consensus.Storage.FS.API (HasFS)
+import           Ouroboros.Consensus.Storage.FS.API.Types (FsPath)
+import           Ouroboros.Consensus.Storage.VolatileDB.API (BlockInfo (..))
+import           Ouroboros.Consensus.Storage.VolatileDB.Impl.Types
+
+{------------------------------------------------------------------------------
+  API
+------------------------------------------------------------------------------}
+
+-- | Parse the given file containing blocks.
+--
+-- Return the 'ParsedBlockInfo' for all the valid blocks in the file. Stop
+-- when encountering an error and include the offset to truncate to.
+newtype Parser m blk = Parser {
+    parse :: FsPath
+          -> m ( [ParsedBlockInfo blk]
+               , Maybe (ParseError blk, BlockOffset)
+               )
+  }
```

Comment:
Hmm, having a separate `Parser` type is convenient. When opening the VolatileDB in `Impl.hs`, we construct the parser and give it to `mkOpenState` in `Impl/State.hs`. Otherwise we have to pass a bunch of arguments to `mkOpenState` and we can't just pass `VolatileDbArgs` because they're defined in `Impl.hs` :slightly_frowning_face:. We could of course move them, but still, not having to pass so many arguments is also convenient. (In `ImmutableDB` we have `ValidateEnv` for that, but there we do a whole lot more validation.)

Reply:
It worked out for the ImmutableDB, though.

Reply:
Done it anyway. Lots of arguments passing and extra constraints, though.